packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -74,324 +186,476 @@ class FocusNode extends ChangeNotifier {
+  ///
+  /// Calling this will clear the [policyData] for the enclosing scope.
+  void requestFocus([FocusNode node]) {
+    if (node != null) {
+      // If an argument was given, then it expects to be a child of this node,
+      // and have the focus given to it.  Reparent the node and then request
+      // focus on it.
+      if (node._parent == null) {
+        reparentIfNeeded(node.context, node);
+      }
+      assert(node.ancestors.contains(this));
+      node.requestFocus();
       return;
-    _focus?.unfocus();
-    node._hasKeyboardToken = true;
-    _setFocus(node);
+    }
+    _doRequestFocus(false);
+  }
+
+  void _doRequestFocus(bool isFromPolicy) {
+    assert(isFromPolicy != null);
+    assert(_manager != null, "Tried to request focus for a node that isn't part of the focus tree.");
+    if (!isFromPolicy) {
+      enclosingScope?.policyData = null;
+    }
+    _hasKeyboardToken = true;
+    _setAsFocusedChild();
+    _manager._markNeedsUpdate(newFocus: this);
   }
 
-  /// If this scope lacks a focus, request that the given node becomes the
-  /// focus.
+  /// Whether this node has focus.
   ///
-  /// Useful for widgets that wish to grab the focus if no other widget already
-  /// has the focus.
+  /// A [FocusNode] has the overall focus when the node is focused in its
+  /// parent [FocusScopeNode] and [FocusScopeNode.hasFocus] is true for
+  /// that scope and all its ancestor scopes.
   ///
-  /// The node is notified that it has received the overall focus in a
-  /// microtask.
-  void autofocus(FocusNode node) {
-    assert(node != null);
-    if (_focus == null) {
-      node._hasKeyboardToken = true;
-      _setFocus(node);
+  /// To request focus, find the [FocusScopeNode] for the current [BuildContext]
+  /// and call the [FocusScopeNode.requestFocus] method:
+  ///
+  /// ```dart
+  /// FocusScope.of(context).requestFocus(focusNode);
+  /// ```
+  ///
+  /// This object notifies its listeners whenever this value changes.
+  bool get hasFocus {
+    if (_manager?._currentFocus == null) {
+      return false;
+    }
+    if (hasPrimaryFocus) {
+      return true;
     }
+    return _manager._currentFocus.ancestors.contains(this);
   }
 
-  /// Adopts the given node if it is focused in another scope.
+  /// Returns true if this node currently has the application-wide focus.
   ///
-  /// A widget that requests that a node is focused should call this method
-  /// during its `build` method in case the widget is moved from one location
-  /// in the tree to another location that has a different focus scope.
-  void reparentIfNeeded(FocusNode node) {
-    assert(node != null);
-    if (node._parent == null || node._parent == this)
-      return;
-    node.unfocus();
-    assert(node._parent == null);
-    if (_focus == null)
-      _setFocus(node);
+  /// This is different from [hasFocus] in that [hasFocus] is true if the node
+  /// is anywhere in the focus chain, but here the node has to be at the end of
```

Comment:
"focus chain" is another important term that would bear explaining in the FocusNode class doc.